### PR TITLE
Fix precedence of !, owned, etc.

### PR DESCRIPTION
**Pull request by vasslitvinov**
Before this change, the parser was treating `!` as the unary-`!` operator.
It had higher precedence than + and lower precedence than : .

This change distinguishes prefix-`!` from postfix-`!`. It moves TBANG
in the precedence table to reflect postfix-`!`. To keep prefix-`!` working
as before, we mark one of its uses with a special `%prec` directive.

This also marks TBORROWED etc. as right associative.
This makes `borrowed borrowed X` parse as borrowed (borrowed X).
Which might not matter in practice.

Meanwhile postfix-`!` and postfix-`?` are left associative.
`x ! !` and `y ? ?` parse as `(x!)!` and `(y?)?` respectively.

Several tests are added to show (some of) these precedences in action.
`bang-vs-new-etc.good` contains lots of misleading lines due to #13946.
The interesting pieces ex. when it is not working as expected are
accompanied with comments in `bang-vs-new-etc.chpl`.

The tests `type-casting.chpl` and `nilable-equalities.chpl` are adjusted
to reflect that `!` now binds more tightly than `owned` et al.

Co-developed with @mppf.

### CODE REVIEW

File: compiler/parser/chapel.ypp [hunk 1]
```diff
@@ -375,12 +375,13 @@
...
-%right TBNOT TBANG
+%right TBNOT TLNOT
 %left TREDUCE TSCAN TDMAPPED
 %right TEXP
 %left TCOLON
-%left TBORROWED TOWNED TUNMANAGED TSHARED
-%right TQUESTION
+%right TBORROWED TOWNED TUNMANAGED TSHARED
+%left TBANG
+%left TQUESTION
```
Comment (mppf) [inline compiler/parser/chapel.ypp]: Since I don't think there is an precedence that can apply between ! and ? I think we should try putting them on the same `%left TQUESTION TBANG` line.

File: test/classes/delete-free/type-casting.chpl [hunk 2]
```diff
@@ -13,7 +13,7 @@ proc makeBorrowedBang1(type t) type {
   return borrowed t!;
 }
 proc makeBorrowedBang2(type t) type {
-  return t:borrowed!;
+  return t:borrowed class;
```
Comment (mppf) [inline test/classes/delete-free/type-casting.chpl]: +1

File: test/classes/delete-free/type-casting.good [hunk 3]
```diff
@@ -0,0 +1,6 @@
+type-casting.chpl:12: In function 'makeBorrowedBang1':
```
Comment (mppf) [inline test/classes/delete-free/type-casting.good]: My intention with this test is that the regular .good file not have any errors in it and any error cases end up in type-casting.errors.good. That means that the `check` calls at line 54 and 60 need to move in to the conditional `if DuplicateManagement` on their next lines.

File: test/classes/nilability/nilable-equalities.chpl [hunk 4]
```diff
@@ -2,11 +2,11 @@
 class MyClass { var x:int; }
 
 type od = owned MyClass;
-type ob = owned MyClass!; // aka borrowed MyClass
+type ob = owned MyClass!; // '!' binds more tightly than 'owned' et al.
```
Comment (mppf) [inline test/classes/nilability/nilable-equalities.chpl]: Please change this to `type ob = (owned MyClass)!` and undo the changes elsewhere in the file

File: test/classes/nilability/nilable-equalities.chpl [hunk 5]
```diff
@@ -2,11 +2,11 @@
...
 type oq = owned MyClass?;
 
 type sd = shared MyClass;
-type sb = shared MyClass!; // aka borrowed MyClass
+type sb = shared MyClass!;
```
Comment (mppf) [inline test/classes/nilability/nilable-equalities.chpl]: Please change this to `type sb = (shared MyClass)!` and undo the other changes.

File: test/parsing/vass/bang-vs-new-etc.chpl [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+proc main {
+  var z01a = new C()?!;
+  check("z01a = new C()?!", z01a, "borrowed C");
+
+  var z01b = new owned C()?!;
+  check("z01b = new owned C()?!", z01b, "borrowed C"); // get owned C
+
+  var z01c = new unmanaged C()?!;  // error: duplicate decorators
```
Comment (mppf) [inline test/parsing/vass/bang-vs-new-etc.chpl]: I don't see any duplicate decorators here, so seems like something to fix in another PR.

File: test/parsing/vass/bang-vs-new-etc.chpl [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+  check("z01c = new unmanaged C()?!", z01c, "unmanaged C");
+
+  var z02 = new C?()!;
+  check("z02 = new C?()!", z02, "borrowed C");
+
+  // var z = new C()!?;  -- error "? on a value"
+  // var z = new C()??; -- syntax error
```
Comment (mppf) [inline test/parsing/vass/bang-vs-new-etc.chpl]: I think both of these have fairly well-defined meanings according to the proposed precedence table.

`new C()!?` returns the same as `new borrowed C?()`
`new C()??` is the same as `new C?()`

For another PR

**Comment by vasslitvinov** [comment]
Follow-up issues: #13950, #13957.
